src/or/hs_service.c

```diff
@@ -1053,6 +1056,222 @@ load_service_keys(hs_service_t *service)
   return ret;
 }
 
+/* Allocate and return a string containing the path to client secret key
+ * in directory. This function will never return NULL. The caller must free
+ * this path. */
+static char *
+client_seckey_path_from_name(const char *directory, const char *client_name)
+{
+  char *file_path = NULL;
+
+  tor_assert(directory);
+  tor_assert(client_name);
+
+  tor_asprintf(&file_path, "%s%s%s%s%s%s", directory,
+                                           PATH_SEPARATOR,
+                                           dir_client_seckeys,
+                                           PATH_SEPARATOR,
+                                           client_name,
+                                           ".privkey");
+  return file_path;
+}
+
+/** Parse the content of a client_key file in <b>client_keys_str</b>
+ * and add hs_service_authorized_client_t's for each parsed client to
+ * <b>parsed_clients</b>. Return the number of parsed clients as result
+ * or -1 for failure. */
+static int
+parse_client_keys(strmap_t *parsed_clients,
+                  const char *start,
+                  const char *end)
+{
```

Comment:
I think it would be much better to use the `smartlist_split_string` model in here like `rend_parse_service_authorization()` does. Each line, we split and extract.